typing/typecore.ml

```diff
@@ -990,13 +968,133 @@ let rec has_literal_pattern p = match p.ppat_desc with
   | Ppat_or (p, q) ->
      has_literal_pattern p || has_literal_pattern q
 
-exception Need_backtrack
-
 let check_scope_escape loc env level ty =
   try Ctype.check_scope_escape env level ty
   with Unify trace ->
     raise(Error(loc, env, Pattern_type_clash(trace, None)))
 
+type pattern_checking_mode =
+  | Normal
+  (** We are checking user code. *)
+
+  | Counter_example of counter_example_checking_info
+  (** We are checking a counter-example to exhaustivity
+      generated by the pattern-matching analyser. *)
+
+
+(** In [Counter_example] mode the responsibility of [type_pat] is to
+   check that a counter-example to exhaustivity is valid: it has to be
+   type-correct in a slightly different sense than for user-code, see
+   [splitting_mode] below, and also capture at least one concrete
+   value.
+
+   [counter_example_checking_info] contains various parameters for
+   this mode of operation. *)
+and counter_example_checking_info = {
+    explode: explosion_fuel;
+    splitting_mode: splitting_mode;
+    synthetic_names: counter_example_synthetic_names;
+  }
+
+
+(** For a pattern to be a valid counter-example, it must be
+   inhabited -- match at least one value. A pattern may fail to match
+   a value if some of its wildcard are at a type that is empty
+   (typically due to incompatible GADT equations, although we also
+   support empty sum types).
+
+   The type-checker tries to figure out if a wildcard is in fact empty
+   by "exploding" it, doing a case-analysis on possible
+   constructors. It only tries up to a fixed depth bound, to avoid
+   non-termination (this is a generic proof-search problem) or
+   compilation-time slowdowns.
+
+   [explosion_fuel] is this depth limit on wildcard explosion. *)
+and explosion_fuel = int
+
+
+(** A counter-example pattern may not be well-typed because the
+   generator does not know about GADTs. In particular, it may generate
+   (even in depth) disjunctions of the form
+
+         (p1 | p2 | ... | pn),
+
+   where some of the pi may in fact be ill-typed. If at least one of
+   them is well-typed, we do have a valid counter-example. So we want
+   the type-checker to succeed if at least one of the alternatives is
+   well-typed.
```

Comment:
That sounds surprising to me.

Again, I'd welcome an actual example. Hopefully you can produce just one example for both comments: an in-depth or-pattern where some of the branches are ill-typed but the counter-example itself accepted.

Reply:
A possible example:
```OCaml
type 'a t = A: 'a -> 'a t |  B : int t
type 'b s = X: int s | Y: float s | Z of 'b
let f: int s t -> _ = fun (A X) -> ()
```
which leads to `type_pat`trying  to type the following counter example: ` B (Y|Z) | A ` where only the patch `B Z (_:int)` is well-typed.